crates/libs/core/src/event.rs

```diff
@@ -25,228 +23,86 @@ impl<T: Interface> Event<T> {
     /// Creates a new, empty `Event<T>`.
     pub fn new() -> Self {
         Self {
-            delegates: Array::new(),
-            swap: Mutex::default(),
-            change: Mutex::default(),
+            delegates: RwLock::new(None),
         }
     }
 
     /// Registers a delegate with the event object.
-    pub fn add(&mut self, delegate: &T) -> Result<i64> {
-        let mut _lock_free_drop = Array::new();
-        Ok({
-            let _change_lock = self.change.lock().unwrap();
-            let mut new_delegates = Array::with_capacity(self.delegates.len() + 1)?;
-            for delegate in self.delegates.as_slice() {
-                new_delegates.push(delegate.clone());
-            }
-            let delegate = Delegate::new(delegate)?;
-            let token = delegate.to_token();
-            new_delegates.push(delegate);
+    pub fn add(&self, delegate: &T) -> Result<i64> {
+        let mut guard = self.delegates.write().unwrap();
+        let new_delegate = Delegate::new(delegate)?;
+        let token = new_delegate.to_token();
+        let new_iter = once(new_delegate);
```

Comment:
```suggestion
        let new_delegate = Delegate::new(delegate)?;
        let token = new_delegate.to_token();
        let new_iter = once(new_delegate);
        let mut guard = self.delegates.write().unwrap();
```

At this point I'm just nitpicking. 😊 The only way this fails if the delegate wrapper cannot be constructed because the delegate cannot be marshaled. So put that first and grab the lock at the last moment.

Reply:
Sure thing 😅